cpp/src/replace/replace.cu

```diff
@@ -87,6 +87,169 @@ __device__ auto get_new_value(cudf::size_type idx,
+                                           cudf::mutable_column_device_view offsets,
+                                           cudf::mutable_column_device_view indices) {
+  cudf::size_type nrows = input.size();
+  cudf::size_type i = blockIdx.x * blockDim.x + threadIdx.x;
+  while (i < nrows) {
+    if (input_has_nulls) {
+      bool input_is_valid = input.is_valid_nocheck(i);
+      if (input_is_valid) {
+        int result = get_new_string_value(i, input, values_to_replace, replacement);
+        if (result == -1) {
+          cudf::string_view output = input.element<cudf::string_view>(i);
+          offsets.data<cudf::size_type>()[i] = output.size_bytes();
+          indices.data<cudf::size_type>()[i] = -1;
+        }
+        else {
+          cudf::string_view output = replacement.element<cudf::string_view> (result);
+          offsets.data<cudf::size_type>()[i] = output.size_bytes();
+          indices.data<cudf::size_type>()[i] = result;
+        }
+      }
+      else {
+        offsets.data<cudf::size_type>()[i] = 0;
+        indices.data<cudf::size_type>()[i] = -1;
+      }
+    }
+    else {
+      int result = get_new_string_value(i, input, values_to_replace, replacement);
+      if (result == -1) {
+        cudf::string_view output = input.element<cudf::string_view>(i);
+        offsets.data<cudf::size_type>()[i] = output.size_bytes();
+        indices.data<cudf::size_type>()[i] = -1;
+      }
+      else {
+        cudf::string_view output = replacement.element<cudf::string_view>(result);
+        offsets.data<cudf::size_type>()[i] = output.size_bytes();
+        indices.data<cudf::size_type>()[i] = result;
+      }
+    }
+    i += blockDim.x * gridDim.x;
+  }
+}
+
+template<bool input_has_nulls, bool replacement_has_nulls>
+__global__ void replace_strings_second_pass(cudf::column_device_view input,
+                                            cudf::column_device_view replacement,
+                                            cudf::mutable_column_device_view offsets,
+                                            cudf::mutable_column_device_view strings,
+                                            cudf::mutable_column_device_view indices,
+                                            cudf::bitmask_type * output_valid,
+                                            cudf::size_type * __restrict__ output_valid_count) {
+  cudf::size_type nrows = input.size();
+  cudf::size_type i = blockIdx.x * blockDim.x + threadIdx.x;
+  uint32_t active_mask = 0xffffffff;
+  active_mask = __ballot_sync(active_mask, i < nrows);
+  auto const lane_id { threadIdx.x % cudf::experimental::detail::warp_size };
+  uint32_t valid_sum { 0 };
+  while (i < nrows) {
+    bool output_is_valid = true;
+    uint32_t bitmask = 0xffffffff;
+
+    if (input_has_nulls) {
+      bool const input_is_valid = input.is_valid_nocheck(i);
+      output_is_valid = input_is_valid;
+      bitmask = __ballot_sync(active_mask, input_is_valid);
+      if (input_is_valid) {
```

Comment:
There's a ton of duplication in this complex chain of nested if/else/elseifs. Factor out the calculation of output and memcpy to reduce duplication.  See my comments on the other kernel for similar guidance.

Reply:
Done.